Launchpad: Add new page task

markbiek opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes https://github.com/Automattic/wp-calypso/issues/77295

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
In the "Keep Building" Launchpad task list, this adds a new `add_new_page` task. The task is marked complete after the user adds their first new page.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply Calypso PR https://github.com/Automattic/wp-calypso/pull/78643 (or use the calypso.live link).
* Apply this patch to Jetpack on your sandbox and sandbox the public-api.
* Go to `/start` in calypso and create a new site with any plan.
* Select the "Promote myself or business goal"
![image](https://github.com/Automattic/jetpack/assets/917632/a4df5373-2f45-47fa-a7a0-effa4fe3f771)
* Launch the site.
* Upon returning to `/home`, you should see a task list like the following which includes the "Add new page" task which is marked incomplete.
![image](https://github.com/Automattic/jetpack/assets/917632/54142482-671f-42d5-ad87-7584da3f9320)
* Click the "Add new page" task. You should be taken to add a new page. Add any kind of page and publish it.
* After returning to `/home`, the "Add new page" task should be crossed off.
![image](https://github.com/Automattic/jetpack/assets/917632/84910849-d4d5-4847-8c2e-6ba6f94699f7)

markbiek committed:
[not verified] New task for adding a page to a site

valterlorran commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -663,6 +682,42 @@ function ( $site_purchase ) {
+ * When a new page is added to the site, mark the add_new_page task complete as needed.
+ *
+ * @param int    $post_id The ID of the post being updated.
+ * @param object $post    The post object.
```
Minor: Shouldn't we use the type `WP_POST` here?

markbiek replied:
We can't do that because it's not PHP 5.6 compatible.

markbiek committed:
[not verified] Fix merge error

markbiek committed:
[not verified] Remove unused variable

sixhours commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -300,6 +307,18 @@ function wpcom_mark_launchpad_task_complete( $task_id ) {
+ * @param string $task_id The task ID.
+ * @return bool True if the task was marked as complete, false otherwise.
+ */
+function wpcom_is_launchpad_task_complete( $task_id ) {
```
I think we already have `wpcom_is_task_option_completed` for this.

sixhours commented:
I created a brand-new site following the testing instructions and the "Add a new page" task was already marked as done when I landed in Customer Home. 🤔 I see you've already accounted for Headstart so I'm not sure why. I noticed the 
homepage is the only page on the site, so maybe it's counting that and it wasn't created/marked as a Headstarted page?

<img width="1072" alt="Screen Shot 2023-06-27 at 4 35 54 PM" src="https://github.com/Automattic/jetpack/assets/2124984/2d47e0ee-4f58-484a-a88f-6de898ae98d8">

Oh, I chose the Side-by-Side theme on signup if that helps; it looks like it's Blank Canvas underneath but allows users to use the pattern assembler to create their own home page. That might explain it!
<img width="504" alt="Screen Shot 2023-06-27 at 4 38 58 PM" src="https://github.com/Automattic/jetpack/assets/2124984/76d45b09-cd00-4f50-9f8b-4c5ae7b37ba0">

markbiek commented:
> I created a brand-new site following the testing instructions and the "Add a new page" task was already marked as done when I landed in Customer Home. 🤔 I see you've already accounted for Headstart so I'm not sure why. I noticed the homepage is the only page on the site, so maybe it's counting that and it wasn't created/marked as a Headstarted page?

Thanks for catching that!

It looks like the Headstart check isn't working. I'm seeing the task get marked as complete as soon as Headstart creates the About page.

markbiek committed:
[not verified] Account for additional default theme content

We were already handling themes that use headstart but we also needed to handle themes that add blocks/pages via the API (like Side-by-Side).

We account for these themes by ignoring pages that are added but not published.

markbiek committed:
[not verified] changelog

markbiek commented:
@sixhours I think this is working now if you'd like to retest.

markbiek committed:
[not verified] Use existing task completion check function

andres-blanco commented:
@markbiek, I didn't dig in to what might be happening here but when I apply this branch on my sandbox and I get this error when hitting the launchpad API. 
<img width="1752" alt="image" src="https://github.com/Automattic/jetpack/assets/375980/ece6451b-7130-4dfe-a217-f1313796c094">
I will dig in further in couple of minutes.

andres-blanco commented on projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php:
```diff
@@ -672,6 +679,46 @@ function ( $site_purchase ) {
+ * @return bool True if add_new_page task is completed.
+ */
+function wpcom_is_add_new_page_completed() {
+	return wpcom_is_task_option_completed( 'add_new_page' );
```
The function `wpcom_is_task_option_completed` recieves an `array` as parameter, not a `string`. https://github.com/Automattic/jetpack/blob/ebf3671dc4f99b9fd0ce44955933fc249fef8e70/projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php#L484

markbiek replied:
Whoops, I'll get that fixed. Thanks!

markbiek committed:
Switch the add_new_page is_complete_callback to use wpcom_is_task_option_completed

markbiek committed:
Correctly use wpcom_is_task_option_completed for manual add_new_page check

andres-blanco reviewed (approved):
LGTM!
